### ISSUE TITLE
affected:apps returning all apps when any changed file is not in any project scope

### ISSUE THREAD

**Issue by mcblum** [nrwl/nx#392]
Sorry for being a pain in the ass. I'm trying to understand exactly how this script works and where it's going wrong for us. I've looked into what's happening and here's what I know:

The script gets a list of changed files, in my case for this test they are:
```
[ 'config/deploy.rb',
  'scripts/build.sh' ]
```

None of these files are related in any way to any of our projects, and I can see that the script knows this because the result of:

```
return touchedFiles.map(function (f) {
        var p = projects.filter(function (project) {
            console.log(project.name, project.files.indexOf(f));
            return project.files.indexOf(f) > -1;
        })[0];
        return p ? p.name : null;
    });
```
logs the name of every project and every instance of `project.files.indexOf(f)` is `-1`. To me that means that it's seeing that there are no projects that are actually affected by any of the changes made between the last commit and this one.

Next, I'm logging within the `affectedApps` function as follow:
```
var tp = touchedProjects(projects, touchedFiles);
console.log('touched projects', tp);
```

and the result of that is:

```
touched projects [ null, null, null, null, null ]
```

The next part of that function is:

```
if (tp.indexOf(null) > -1) {
        return projects.filter(function (p) { return p.type === ProjectType.app; }).map(function (p) { return p.name; });
    }
```

which is obviously true in our case, which then returns the names of every app we have. That's what I can't quite figure out - if the results of the first query where the script checks if the touched files relate to a project is `-1`, and the result of touchedProjects is `null` for each project, then returning the name of every project seems like the opposite of what I would expect. 

Our desired result is that if no projects are affected by the changes between two SHAs, then nothing is built. 

I'm sure I'm just missing something here. Once again thank you for your work on Nx.

**Comment by mcblum** [comment]
The more I stare at this the more it feels like maybe the reason for this behavior is that if the file that is changed is completely outside the scope of every application, one could assume that it may have a global impact? In this case I've modified a couple build scripts which does affect everything so, in a way, even though the result from the script isn't what I expected it does sort of make sense.

I'm not sure if we want to rebuild every application every time a file is changed outside of the projects' scopes, so I might modify the script to accept another argument such a `--allow-null` or something that indicates that if the files do not affect any project that it's ok to return nothing. 

I'm not confident that I'm right about how everything is working, so there could be unintended consequences for this.

**Comment by dereklin** [comment]
Ideally the tool can detect the truly affected apps.  However, that might be very difficult to do.  It's better to error on the safe side.

**Comment by mcblum** [comment]
@dereklin totally - and I totally get that. The only issue that we're seeing is that if anyone modifies a file outside the scope of all applications it will trigger a build that takes around 50 minutes now. I guess it's not the end of the world, servers do their thing, but it also seems like a lot of compute cycles.

**Comment by dereklin** [comment]
For the files that won't affect any app but are in the source code?  Are they being modified pretty frequently?

If so, maybe nx can provide some ignore configuration file.

**Comment by mcblum** [comment]
@dereklin it depends if we're adding apps or working on dev environments or the pipeline. This time I added an app and because of that modified a couple scripts. But I totally agree, an `.nxignore` would be a great idea. I think I'll add that to my open pull request and see if they are in to it.

**Comment by mrmeku** [comment]
Can you associate your open pull request with this bug? You can do so by changing your PR description to reference this bug. Also, your PR doesn't conform to our naming conventions defined in https://github.com/nrwl/nx/blob/master/CONTRIBUTING.md . If you could update the PR to meet those conventions that'd be great

**Comment by mcblum** [comment]
@mrmeku absolutely - let me fix the auto-formatting that my IDE did as well. thank you!

**Comment by mcblum** [comment]
@mrmeku ok, I think I got that right. to be honest I think this is my second pull request ever for a public repo so please let me know if I screwed anything up.

**Comment by vsavkin** [comment]
Closing this as the issue has been addressed in Nx 6.1. The nx.json file has a property called `implicitDependencies`. This property specified how "global" files affect projects. You can read more about it here: https://blog.nrwl.io/nrwl-nx-6-1-better-dev-ergonomics-faster-builds-3198bb310e39
